### PR TITLE
Add syntactic sugar for attr.ib(default=attr.Factory)

### LINKED ISSUE #178: attr.Factory is a little wordy
attrs is a huge net win on an object like this:

```python
@attr.s
class Point(object):
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

import attr
@attr.s
class Point(object):
    x = attr.ib()
    y = attr.ib()
    z = attr.ib()
```

but it's a lot less clear when you have something like this:

```python
class Cache(object):
    def __init__(self):
        self._stored = []
        self._by_name = {}
        self._by_id = {}
```

which becomes

```python
@attr.s
class Cache(object):
    _stored = attr.ib(default=attr.Factory(list))
    _by_name = attr.ib(default=attr.Factory(dict))
    _by_id = attr.ib(default=attr.Factory(dict))
```

I think an alias for this behavior, like:

```python
@attr.s
class Cache(object):
    _stored = attr.ib(new=list)
    _by_name = attr.ib(new=dict)
    _by_id = attr.ib(new=dict)
```

could make initializing these types of mutable objects a lot less verbose.

### PR DESCRIPTION
**Pull request by hynek**
Let's help @glyph to type less.

Fixes #178

### CODE REVIEW

File: src/attr/_make.py [hunk 1]
```diff
@@ -156,6 +162,14 @@ def attrib(default=NOTHING, validator=None,
         )
         converter = convert
 
+    if factory is not None:
+        if default is not NOTHING:
+            raise ValueError(
+                "The `default` and `factory` arguments are mutually "
+                "exclusive."
+            )
+        default = Factory(factory)
```
Comment (RonnyPfannschmidt) [inline src/attr/_make.py]: for sanity, can factory fail if its being passed something other than a callable?
i can already see people writing `factory=Factory(...)`

File: tests/test_make.py [hunk 2]
```diff
@@ -546,6 +546,18 @@ class C(object):
 
         assert "mutually exclusive" in ei.value.args[0]
 
+    def test_sugar_callable(self):
+        """
+        Factory has to be a callable to prevent people from passing Factory
+        into it.
+        """
+        with pytest.raises(ValueError) as ei:
```
Comment (RonnyPfannschmidt) [inline tests/test_make.py]: `with pytest.raises(ValueError, match="must be a callable"): ...` might make this more pretty

Comment (hynek) [reply tests/test_make.py]: ohhh…shiny new pytest feature or am I just out of touch? :)

Comment (RonnyPfannschmidt) [reply tests/test_make.py]: https://github.com/pytest-dev/pytest/pull/2227 was merged a year ago, but its a detail easy to miss

**Comment by hynek** [comment]
@chadrik I guess this needs into the stubs? 😇

**Comment by euresti** [comment]
Yeah it will.  I'm trying to figure out the best time to update the stubs?  When is the next release happening?

**Comment by hynek** [comment]
I don’t know could be any time but I don’t really have any pressure. Any wishes?

**Comment by euresti** [comment]
I think my preference would be to update all the stubs at once per release.  Sadly you can't version the stubs per library version so I don't want to give people old stubs.  Although I don't know how often mypy releases come.  

Oh bother. I can't wait for PEP 561 so we can just include the stubs with the library and be perfectly happy about it.

Oh I just noticed that this feature means I have to update the plugin too.  So that the __init__ has a default.  

Anyway when is the next release planned for?

**Comment by hynek** [comment]
As said: I don’t have a schedule, any wishes? Should be March tho. ;)

**Comment by chadrik** [comment]
Luckily my tests on the pyi_stubs branch will flag these issues when it gets merged :)

> I can't wait for PEP 561, so we can just include the stubs with the library and be perfectly happy about it.

It's getting close: https://github.com/python/mypy/pull/4693

However, after it lands keeping the plugin and stubs in sync is going to be a bit of a pain, especially considering my [PR for pip-installable mypy plugins](https://github.com/python/mypy/pull/4734) was shot down for now.  For example, I'm assuming this particular feature needs to be supported by the stubs and the plugin in tandem...  what happens when they're out of sync with each other?  At least with the typeshed we can ensure that each mypy release contains a compatible version of the plugin and stubs, even though it's only for bleeding edge versions of attrs.  

Once PEP561 lands, and while we're still waiting on pip-installable plugins, I think each release of attrs will need to specify a compatible mypy version.

**Comment by euresti** [comment]
Oops. Misread the message.  I might have time this weekend to work on the plugin.   But now that I think about it, since it depends on mypy shipping a release (which won't happen for a while) there's no real rush here.  People who want to use this feature and type-check will probably have to install an unreleased version of mypy.

**Comment by hynek** [comment]
I guess it would be good to have #238 in 18.1 tho?

**Comment by euresti** [comment]
Ok I have both a typeshed diff and a mypy diff ready to support this.  I'd like to do a single "support attrs 18.1" PR are there any other features?    I got fields_dict also.

https://github.com/euresti/mypy/commit/ee19aedc2fb8dd9f9d2268c0b04008b78ac712a4
https://github.com/euresti/typeshed/commit/20e3cf866357577ff5dac40670f45fb5455fd69f

**Pull request merged**